test/benchmark/cuda_driver.cpp

```diff
@@ -0,0 +1,197 @@
+#include "llvm/Target/TargetMachine.h"
+
+using fmt::literals::operator""_format;
+
+namespace nmodl {
+namespace runner {
+
+void CUDADriver::checkCudaErrors(CUresult err) {
+    if (err != CUDA_SUCCESS) {
+        const char* ret = NULL;
+        cuGetErrorName(err, &ret);
+        throw std::runtime_error("CUDA error: " + std::string(ret));
+    }
+}
+
+void CUDADriver::link_libraries(llvm::Module& module, BenchmarkInfo* benchmark_info) {
+    llvm::Linker linker(module);
+    for (const auto& lib_path: benchmark_info->shared_lib_paths) {
+        const auto lib_name = lib_path.substr(lib_path.find_last_of("/\\") + 1);
+        std::regex libdevice_bitcode_name{"libdevice.*.bc"};
+        if (!std::regex_match(lib_name, libdevice_bitcode_name)) {
+            throw std::runtime_error("Only libdevice is supported for now");
+        }
+        // Load libdevice module to the LLVM Module
+        auto libdevice_file_memory_buffer = llvm::MemoryBuffer::getFile(lib_path);
+        llvm::Expected<std::unique_ptr<llvm::Module>> libdevice_expected_module =
+            parseBitcodeFile(libdevice_file_memory_buffer->get()->getMemBufferRef(),
+                             module.getContext());
+        if (std::error_code error = errorToErrorCode(libdevice_expected_module.takeError())) {
+            throw std::runtime_error("Error reading bitcode: {}"_format(error.message()));
+        }
+        linker.linkInModule(std::move(libdevice_expected_module.get()),
+                            llvm::Linker::LinkOnlyNeeded);
+    }
+}
+
+void print_string_to_file(const std::string& ptx_compiled_module, const std::string& filename) {
+    std::ofstream ptx_file(filename);
+    ptx_file << ptx_compiled_module;
+    ptx_file.close();
+}
+
+CUjit_target get_compute_architecture(const int compute_version_major,
+                                      const int compute_version_minor) {
+    auto compute_architecture = compute_version_major * 10 + compute_version_minor;
+    switch (compute_architecture) {
+    case 20:
+        return CU_TARGET_COMPUTE_20;
+    case 21:
+        return CU_TARGET_COMPUTE_21;
+    case 30:
+        return CU_TARGET_COMPUTE_30;
+    case 32:
+        return CU_TARGET_COMPUTE_32;
+    case 35:
+        return CU_TARGET_COMPUTE_35;
+    case 37:
+        return CU_TARGET_COMPUTE_37;
+    case 50:
+        return CU_TARGET_COMPUTE_50;
+    case 52:
+        return CU_TARGET_COMPUTE_52;
+    case 53:
+        return CU_TARGET_COMPUTE_53;
+    case 60:
+        return CU_TARGET_COMPUTE_60;
+    case 61:
+        return CU_TARGET_COMPUTE_61;
+    case 62:
+        return CU_TARGET_COMPUTE_62;
+    case 70:
+        return CU_TARGET_COMPUTE_70;
+    case 72:
+        return CU_TARGET_COMPUTE_72;
+    case 75:
+        return CU_TARGET_COMPUTE_75;
+    case 80:
+        return CU_TARGET_COMPUTE_80;
+    case 86:
+        return CU_TARGET_COMPUTE_86;
+    default:
+        throw std::runtime_error("Unsupported compute architecture");
```

Comment:
I probably don't know full workflow here but one comment come to my mind: Newer GPUs can run binaries compiled for older generations. If detected compute capability is "bigger", do we need to throw an exception or we can get away by targeting older architecture?

Reply:
There is no need to worry about older/newer generations here since for the CUDA JIT we actually read the exact architecture of the GPU we're going to use and pass it to it. This function just makes the conversion from the device target architecture of the exact GPU we are using to `CUjit_target` to be used by the CUDA JIT. I should rename the name of the function though to make it more clear

Reply:
sounds good!